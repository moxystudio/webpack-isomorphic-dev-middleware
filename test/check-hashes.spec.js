'use strict';

const express = require('express');
const request = require('supertest');
const webpackIsomorphicDevMiddleware = require('../');
const createCompiler = require('./util/createCompiler');
const createCompilation = require('./util/createCompilation');
const createWritter = require('./util/createWritter');
const configClientBasic = require('./configs/client-basic');
const configServerBasic = require('./configs/server-basic');
const configServerWithHashes = require('./configs/server-with-hashes');

afterEach(() => jest.restoreAllMocks());
afterEach(() => createCompiler.teardown());

it('should warn if hashes are being used in webpack config', async () => {
    const app = express();
    const compiler = createCompiler(configClientBasic, configServerWithHashes);
    const writter = createWritter();

    app.use(webpackIsomorphicDevMiddleware(compiler, {
        report: { write: writter },
    }));

    await request(app)
    .get('/client.js')
    .expect(200);

    expect(writter.getReportOutput()).toMatchSnapshot();
});

it('should warn if hashes are being used in webpack config (rebuild)', () => {
    const app = express();
    const compiler = createCompiler(configClientBasic, configServerBasic);
    const writter = createWritter();

    compiler.watch = () => {};

    app.use(webpackIsomorphicDevMiddleware(compiler, {
        report: { write: writter },
    }));

    compiler.emit('begin');
    compiler.emit('end', createCompilation());
    compiler.emit('begin');
    compiler.emit('end', createCompilation({
        clientStats: {
            toJson: () => ({
                assets: [{ name: 'foo.a1aaaaaaaaaaaaaaaaaaaaaaa.js' }],
            }),
        },
    }));

    expect(writter.getReportOutput()).toMatchSnapshot();
});

it('should not warn about hashes are being used in webpack config if options.memoryFs is false', async () => {
    const app = express();
    const compiler = createCompiler(configClientBasic, configServerWithHashes);
    const writter = createWritter();

    app.use(webpackIsomorphicDevMiddleware(compiler, {
        memoryFs: false,
        report: { write: writter },
    }));

    await request(app)
    .get('/client.js')
    .expect(200);

    expect(writter.getReportOutput()).toMatchSnapshot();
});

it('should fallback to `process.stderr.write` when printing the warnings', async () => {
    const app = express();
    const compiler = createCompiler(configClientBasic, configServerWithHashes);

    jest.spyOn(process.stderr, 'write');
    app.use(webpackIsomorphicDevMiddleware(compiler, {
        report: false,
    }));

    await request(app)
    .get('/client.js')
    .expect(200);

    expect(process.stderr.write).toHaveBeenCalledTimes(1);
    expect(process.stderr.write.mock.calls[0]).toMatchSnapshot();
});
