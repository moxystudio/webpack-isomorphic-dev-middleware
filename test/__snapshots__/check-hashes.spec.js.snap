// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fallback to \`process.stderr.write\` when printing the warnings 1`] = `
Array [
  "[33mWARN[39m: Assets with an hash in its name were detected on the [1mserver[22m:
- output.a1aaaaaaaaaaaaaaaaaaaaaaa.js

This is known to cause [1mmemory leaks[22m with [1mwebpack-dev-middleware's[22m in-memory filesystem.
You should avoid using [1m[hash][22m in output.filename, output.chunkFilename, output.hotUpdateMainFilename, output.hotUpdateChunkFilename as well as similar options in loaders & plugins.
Alternatively, you may set \`memoryFs\` to false altough it will still create many files in the output folder.
If you feel this was a false positive, please ignore this warning.

",
]
`;

exports[`should not warn about hashes are being used in webpack config if options.memoryFs is false 1`] = `
"● Compiling...
✔ Compilation succeeded (xxx)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    client.js  xxx

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    output.a1aaaaaaaaaaaaaaaaaaaaaaa.js  xxx

"
`;

exports[`should warn if hashes are being used in webpack config (rebuild) 1`] = `
"● Compiling...
✔ Compilation succeeded (xxx)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size foo.js 10Kb

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size bar.js 10Kb

● Compiling...
✔ Compilation succeeded (xxx)
WARN: Assets with an hash in its name were detected on the client:
- foo.a1aaaaaaaaaaaaaaaaaaaaaaa.js

This is known to cause memory leaks with webpack-dev-middleware's in-memory filesystem.
You should avoid using [hash] in output.filename, output.chunkFilename, output.hotUpdateMainFilename, output.hotUpdateChunkFilename as well as similar options in loaders & plugins.
Alternatively, you may set \`memoryFs\` to false altough it will still create many files in the output folder.
If you feel this was a false positive, please ignore this warning.

● Compiling...
✔ Compilation succeeded (xxx)
"
`;

exports[`should warn if hashes are being used in webpack config 1`] = `
"● Compiling...
✔ Compilation succeeded (xxx)

     CLIENT                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    client.js  xxx

     SERVER                             
    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
    Asset Size Chunks Chunk Names
    output.a1aaaaaaaaaaaaaaaaaaaaaaa.js  xxx

WARN: Assets with an hash in its name were detected on the server:
- output.a1aaaaaaaaaaaaaaaaaaaaaaa.js

This is known to cause memory leaks with webpack-dev-middleware's in-memory filesystem.
You should avoid using [hash] in output.filename, output.chunkFilename, output.hotUpdateMainFilename, output.hotUpdateChunkFilename as well as similar options in loaders & plugins.
Alternatively, you may set \`memoryFs\` to false altough it will still create many files in the output folder.
If you feel this was a false positive, please ignore this warning.

"
`;
